### PR TITLE
crypto: Add atomic variant for CC310 locking

### PR DESCRIPTION
**Pull request by lmaciejonczyk**
New lock variant fasten mutually exclusive access to CC3XX resources at the expense of a need to handle potential access failure
in the upper layer.

needed by: https://github.com/nrfconnect/sdk-nrf/pull/3339

### CODE REVIEW

File: crypto/Kconfig
Comment (frkv) [inline crypto/Kconfig:44]: Enables exclusive access to the Arm CryptoCell HW using an RTOS mutex that has priority inheritance. The mutex lock is safe across threads and/or interrupts with different priority. Using an RTOS mutex comes at a cost.

File: crypto/Kconfig
Comment (edmont) [inline crypto/Kconfig:44]:
```suggestion
	  Enables exclusive access to the ARM CryptoCell hardware using an RTOS mutex
```

File: crypto/Kconfig
Comment (edmont) [inline crypto/Kconfig:47]: Be more specific here? Which cost?

File: crypto/Kconfig
Comment (edmont) [inline crypto/Kconfig:55]: Failure accesses to the resource...

**Review by edmont — CHANGES_REQUESTED** [review]
Some editorial suggestions. But please check indentation, it's not uniform across the files.

File: crypto/Kconfig
Comment (lmaciejonczyk) [reply crypto/Kconfig:55]: @b-gent can you help refactor this sentence to be more readable?

File: crypto/Kconfig
Comment (lmaciejonczyk) [reply crypto/Kconfig:47]:
```suggestion
	  at a cost of lower performance.
```
?

Comment (edmont) [reply crypto/Kconfig:47]: Okay.

File: crypto/Kconfig
Comment (b-gent) [reply crypto/Kconfig:55]: Note that if different threads try to get the same resource at the same time, the access will fail. This situation must be resolved by higher application layer.

@lmaciejonczyk  does that work?

File: crypto/Kconfig
Comment (lmaciejonczyk) [inline crypto/Kconfig:57]:
```suggestion
	  frequency requests from one thread at the same time.
	  Note that if different threads try to get the same resource at the same time, the access will fail.
	  This situation must be resolved by higher application layer.
```

**Review by b-gent — APPROVED** [review]

**Review by edmont — CHANGES_REQUESTED** [review]
LGTM, but please replace tabs by four spaces.

**Comment by lmaciejonczyk** [comment]
> LGTM, but please replace tabs by four spaces.

Why? Tabs are commonly used in Kconfig files over all projects.

**Comment by lmaciejonczyk** [comment]
@frkv If you are ok with the changes approve this PR so that we can proceed with merging.

**Comment by edmont** [comment]
> Why? Tabs are commonly used in Kconfig files over all projects.

I mean in the C files. Currently there are mixed styles within the same files.

**Comment by lmaciejonczyk** [comment]
> > Why? Tabs are commonly used in Kconfig files over all projects.
> 
> I mean in the C files. Currently there are mixed styles within the same files.

It's done. Check and approve if it's ok now please.

**Review by edmont — APPROVED** [review]

File: crypto/nrf_cc310_platform/src/nrf_cc3xx_platform_mutex_zephyr.c
Comment (frkv) [inline crypto/nrf_cc310_platform/src/nrf_cc3xx_platform_mutex_zephyr.c:254]: This will still need to add the test for uninitialized mutexes (i.e. mutexes in contexts that enable their own mutex because of MBEDTLS_THREADING_C)

_Note that the scheme for the flags should probably be "is_set(SOME_FLAG)" instead of == since we will need to add another flag type to indicate distinction between atomic mutex and regular kernel mutex (see different comment)_

File: crypto/nrf_cc310_platform/src/nrf_cc3xx_platform_mutex_zephyr.c
Comment (frkv) [inline crypto/nrf_cc310_platform/src/nrf_cc3xx_platform_mutex_zephyr.c:235]: The rng_mutex is very difficult to control since it can be called by anything in Zephyr that needs entropy. This merits handing RNG in a special way. Please change rng_mutex back to use a kernel mutex and handle the additional type difference with a new flag in nrf_cc3xx_platform_mutex.h

_The new flag could be called NRF_CC3XX_PLATFORM_MUTEX_MASK_IS_ATOMIC, so there is a way to distinguish between regular kernel mutex and atomic mutex_

**Review by frkv — CHANGES_REQUESTED** [review]
Requests needs to be addressed before PR can be approved

File: crypto/Kconfig
Comment (frkv) [inline crypto/Kconfig:52]: Using atomic operations is the fastest way to ensure mutually exclusive access to the ARM CryptoCell hardware.

Warning: If this configuration is set, every execution requiring use of the ARM CryptoCell hardware must happen in the same priority. Calling into mbed TLS APIs from a higher priority while an existing operation leads to undefined behavior.  It is highy recommended to to do all crypto operations in one single thread if this configuration is set.

**Review by frkv — APPROVED** [review]
LGTM

**Comment by tejlmand** [comment]
where is the manifest PR that references this PR ?

**Comment by MarekPorwisz** [comment]
I've cherry-picked and added changes from https://github.com/nrfconnect/sdk-nrfxlib/pull/350 here

**Comment by tejlmand** [comment]
@lmaciejonczyk could you please check the compliance failures, seems to be coding violations.

**Comment by lmaciejonczyk** [comment]
> @lmaciejonczyk could you please check the compliance failures, seems to be coding violations.

How can I identify it?

**Review by tejlmand — APPROVED** [review]

**Pull request merged**
